From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 25 Dec 2021 16:39:52 +1000
Subject: [PATCH] Reobf error with ServerPlaceRecipe


diff --git a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
index 80f032cca3309b709a11898cbc08bce27cbd624d..bebb6e739384bf7d314b45b1e04b244cf80fea8f 100644
--- a/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
+++ b/src/main/java/net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -206,7 +206,7 @@ public class ServerPlaceRecipe<C extends Container> implements PlaceRecipe<Integ
     private int getAmountOfFreeSlotsInInventory() {
         int i = 0;
 
-        for (ItemStack itemStack : this.inventory.items) {
+        for (ItemStack itemStack : this.inventory.items) { // MultiPaper fix Reobf error
             if (itemStack.isEmpty()) {
                 i++;
             }
