From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 6 Jul 2022 16:15:42 +1000
Subject: [PATCH] Sync FireworkRocket's attachedToEntity


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index 813c4a52c8ddd70e252125ebfd3654346c71f13c..7939bdfe3d02f3a399972b9a07a359898786d32e 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -321,6 +321,11 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
             nbt.putUUID("SpawningEntity", this.spawningEntity);
         }
         // Paper end
+        // MultiPaper start
+        if (this.attachedToEntity != null) {
+            nbt.putUUID("AttachedToEntity", this.attachedToEntity.getUUID());
+        }
+        // MultiPaper end
     }
 
     @Override
@@ -342,6 +347,11 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
             this.spawningEntity = nbt.getUUID("SpawningEntity");
         }
         // Paper end
+        // MultiPaper start
+        if (nbt.hasUUID("AttachedToEntity") && this.level().getEntities().get(nbt.getUUID("AttachedToEntity")) instanceof LivingEntity livingEntity) {
+            this.attachedToEntity = livingEntity;
+        }
+        // MultiPaper end
     }
 
     @Override
